spec/requests/twitch_stream_updates_spec.rb

```diff
@@ -0,0 +1,127 @@
+        get "/users/#{user.id}/twitch_stream_updates", params: twitch_webhook_subscription_params
+
+        expect(response.status).to eq 204
+      end
+    end
+  end
+
+  describe "POST /users/:user_id/twitch_stream_updates" do
+    context "when the user was not streaming and starts streaming" do
+      let(:currently_streaming_on) { nil }
+
+      let(:twitch_webhook_params) do
+        {
+          data: [{
+            id: "0123456789",
+            user_id: "5678",
+            user_name: "wjdtkdqhs",
+            game_id: "21779",
+            community_ids: [],
+            type: "live",
+            title: "Best Stream Ever",
+            viewer_count: 417,
+            started_at: "2017-12-01T10:09:45Z",
+            language: "en",
+            thumbnail_url: "https://link/to/thumbnail.jpg"
+          }]
+        }
+      end
+      let(:twitch_webhook_secret_sha) do
+        digest = Digest::SHA256.new
+        digest << ApplicationConfig["TWITCH_WEBHOOK_SECRET"]
+        digest << twitch_webhook_params.to_json
+        "sha256=#{digest.hexdigest}"
+      end
+
+      it "updates the Users twitch streaming status" do
+        expect { post "/users/#{user.id}/twitch_stream_updates", params: twitch_webhook_params.to_json, headers: { "Content-Type" => "application/json", "X-Hub-Signature" => twitch_webhook_secret_sha } }.
+          to change { user.reload.currently_streaming? }.from(false).to(true).
+          and change { user.reload.currently_streaming_on_twitch? }.from(false).to(true)
+      end
+    end
+
+    context "when the webhook secret was NOT verified" do
+      let(:twitch_webhook_params) do
+        {
+          data: [{
+            id: "0123456789",
+            user_id: "5678",
+            user_name: "wjdtkdqhs",
+            game_id: "21779",
+            community_ids: [],
+            type: "live",
+            title: "Best Stream Ever",
+            viewer_count: 417,
+            started_at: "2017-12-01T10:09:45Z",
+            language: "en",
+            thumbnail_url: "https://link/to/thumbnail.jpg"
+          }]
+        }
+      end
+      let(:twitch_webhook_secret_sha) { "sha256=BAD_HASH" }
+
+      it "noops" do
+        expect { post "/users/#{user.id}/twitch_stream_updates", params: twitch_webhook_params.to_json, headers: { "Content-Type" => "application/json", "X-Hub-Signature" => twitch_webhook_secret_sha } }.
+          to not_change { user.reload.currently_streaming? }.from(false).
+          and not_change { user.reload.currently_streaming_on_twitch? }.from(false)
+      end
+    end
+  end
+
+  context "when the user was streaming and stops" do
+    let(:currently_streaming_on) { :twitch }
+
+    let(:twitch_webhook_params) do
+      {
+        data: []
+      }
+    end
+    let(:twitch_webhook_secret_sha) do
+      digest = Digest::SHA256.new
+      digest << ApplicationConfig["TWITCH_WEBHOOK_SECRET"]
+      digest << twitch_webhook_params.to_json
+      "sha256=#{digest.hexdigest}"
```

Comment:
This is literally the same lines of code that I am using to verify the subscription. So I don't really think we are testing a ton here... But not sure of a better way to do this

Reply:
you could cheat a bit, create a sha with the code you know it works and make sure the `expect` checks that sha, so that if I go and change the algorithm the digest will not correspond and the test will fail. It's not foolproof but it's something